packages/developer-portal/src/components/apps/pipeline/pipeline-environment.tsx

```diff
@@ -0,0 +1,188 @@
+  const [isInserting, setIsInserting] = useState<boolean>(false)
+  const [modalOpen, setModalOpen] = useState<string | false>(false)
+  const [keys, setKeys] = useState<string[]>([])
+
+  const [fetchedKeys, isFetching] = useReapitGet<string[]>({
+    reapitConnectBrowserSession,
+    action: getActions(window.reapit.config.appEnv)[GetActionNames.getPipelineEnvironment],
+    uriParams: {
+      pipelineId: appId,
+    },
+    headers: {
+      Authorization: connectSession?.idToken as string,
+    },
+    fetchWhenTrue: [connectSession?.idToken],
+  })
+
+  const [sending, , func] = useReapitUpdate<{ key: string; value: string; existingKeys: string[] }, void>({
+    reapitConnectBrowserSession,
+    action: updateActions(window.reapit.config.appEnv)[UpdateActionNames.upsertPipelineEnvironment],
+    uriParams: {
+      pipelineId: appId,
+    },
+    method: 'PUT',
+    headers: {
+      Authorization: connectSession?.idToken as string,
+    },
+  })
+
+  const {
+    register,
+    handleSubmit,
+    formState: { errors },
+    setValue,
+    resetField,
+  } = useForm<{ key: string; value: any; existingKeys: string[] }>({
+    resolver: yupResolver(schema),
+    defaultValues: {
+      existingKeys: [],
+    },
+  })
+
+  useEffect(() => {
+    if (fetchedKeys) {
+      setKeys(fetchedKeys)
+
+      setValue('existingKeys', fetchedKeys)
+    }
+  }, [fetchedKeys])
+
+  const submitParameter = handleSubmit(async (values) => {
+    const result = await func(values)
+    if (result) {
+      resetField('key')
+      resetField('value')
+      setIsInserting(false)
+      setKeys([...keys, values.key])
+    }
+  })
+
+  return (
+    <>
+      <PipelineTabs />
+      <BodyText hasGreyText hasSectionMargin>
+        Here you can add your Environment variables for your application.
+      </BodyText>
+      <div className={cx(elMb6, elMt6)}>
+        <Table>
+          <TableHeadersRow>
+            <TableHeader>Key</TableHeader>
+            <TableHeader></TableHeader>
+          </TableHeadersRow>
+          {isFetching ? (
+            <TableRow>
+              <TableCell>
+                <Loader />
+              </TableCell>
+            </TableRow>
+          ) : (
+            <>
+              {(keys || []).map((key) => (
+                <TableRow key={key}>
+                  <TableCell>{key}</TableCell>
+                  <TableCell>
+                    <Button
+                      intent="secondary"
+                      onClick={() => {
+                        resetField('key')
+                        resetField('value')
+                        setModalOpen(key)
+                      }}
+                    >
+                      Update
+                    </Button>
+                  </TableCell>
+                </TableRow>
+              ))}
+            </>
+          )}
+        </Table>
+      </div>
+      <Modal isOpen={isInserting} onModalClose={() => setIsInserting(false)} className={cx(elMb6, elMt6)}>
```

Comment:
So the pattern for forms with tables is to put the form in the slide down component. Forms in Modals, generally Ana regards as a pattern to avoid if poss. If you use the Table React shorthand, it make this process much easier.